cyrus/syslog/syslog.3
.\" $Revision: 1.1 $
.\" Copyright (c) 1985 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms are permitted provided
.\" that: (1) source distributions retain this entire copyright notice and
.\" comment, and (2) distributions including binaries display the following
.\" acknowledgement:  ``This product includes software developed by the
.\" University of California, Berkeley and its contributors'' in the
.\" documentation or other materials provided with the distribution and in
.\" all advertising materials mentioning features or use of this software.
.\" Neither the name of the University nor the names of its contributors may
.\" be used to endorse or promote products derived from this software without
.\" specific prior written permission.
.\" THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
.\" WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\"	@(#)syslog.3	6.13 (Berkeley) 6/23/90
.\"
.TH SYSLOG 3 "June 23, 1990"
.UC 5
.SH NAME
syslog, openlog, closelog, setlogmask, vsyslog \- control system log
.SH SYNOPSIS
.B "#include <syslog.h>
.PP
.B "openlog(ident, logopt, facility)
.br
.B "char *ident;
.PP
.B "syslog(priority, message, parameters ... )
.br
.B "int priority;
.br
.B "char *message;
.PP
.B "#include <varargs.h>
.br
.B "vsyslog(priority, message, args)
.br
.B "int priority;
.br
.B "char *message;
.br
.B "va_list args;
.PP
.B "closelog()
.PP
.B "setlogmask(maskpri)
.SH DESCRIPTION
.I Syslog
arranges to write
.I message
onto the system log maintained by
.IR syslogd (8).
The message is tagged with
.IR priority .
The message looks like a
.IR printf (3)
string except that
.B %m
is replaced by the current error message (as referenced by
.IR errno ).
A trailing newline is added if needed.
An alternate form, in which the arguments have already been captured
using the variable-length argument facilities of
.IR varargs (3),
is available under the name
.IR vsyslog .
.PP
This message will be read by
.IR syslogd (8)
and written to the system console, log files, or forwarded to
.I syslogd
on another host as appropriate.
.PP
Priorities are encoded as a
.I facility
and a
.IR level .
The facility describes the part of the system
generating the message.
The level is selected from an ordered list:
.TP
LOG_ALERT
A condition that should be corrected immediately,
such as a corrupted system database.
.TP
LOG_CRIT
Critical conditions,
e.g., hard device errors.
.TP
LOG_DEBUG
Messages that contain information
normally of use only when debugging a program.
.TP
LOG_EMERG
A panic condition.
This is normally broadcast to all users.
.TP
LOG_ERR
Errors.
.TP
LOG_INFO
Informational messages.
.TP
LOG_NOTICE
Conditions that are not error conditions,
but should possibly be handled specially.
.TP
LOG_WARNING
Warning messages.
.PP
If
.I syslog
cannot pass the message to
.IR syslogd ,
it will attempt to write the message to the console (``/dev/console'')
if the LOG_CONS option is set (see below).
.PP
If special processing is needed,
.I openlog
can be called to initialize the log file.
The parameter
.I ident
is a string that is prepended to every message.
.I Logopt
is a bit field indicating logging options.
Current values for
.I logopt
are:
.TP
LOG_CONS
If unable to send the message to
.IR syslogd ,
write it to the console.
.TP
LOG_NDELAY 
Open the connection to
.I syslogd
immediately.
Normally the open is delayed
until the first message is logged.
Useful for programs that need to manage the
order in which file descriptors are allocated.
.TP
LOG_PERROR
Write the message to stderr as well to the system log.
.TP
LOG_PID
Log the process id with each message:
useful for identifying instantiations of daemons.
.PP
The
.I facility
parameter encodes a default facility to be assigned to all messages
that do not have an explicit facility encoded:
.TP
LOG_AUTH
The authorization system:
.IR login (1),
.IR su (1),
.IR getty (8),
etc.
.TP
LOG_CRON
The clock daemon.
.TP
LOG_DAEMON
System daemons, such as
.IR ftpd (8),
.IR routed (8),
etc, that are not provided for explicitly by other facilities.
.TP
LOG_KERN
Messages generated by the kernel.
These cannot be generated by any user processes.
.TP
LOG_LPR
The line printer spooling system:
.IR lpr (1),
.IR lpc (8),
.IR lpd (8),
etc.
.TP
LOG_MAIL
The mail system.
.TP
LOG_NEWS
The network news system.
.TP
LOG_SYSLOG
Messages generated internally by
.IR syslogd (8).
.TP
LOG_USER
Messages generated by random user processes.
This is the default facility identifier if none is specified.
.TP
LOG_UUCP
The uucp system.
.TP
LOG_LOCAL0
Reserved for local use.
Similarly for LOG_LOCAL1 through LOG_LOCAL7.
.PP
.I Closelog
can be used to close the log file.
.PP
.I Setlogmask
sets the log priority mask to
.I maskpri
and returns the previous mask.
Calls to
.I syslog
with a priority not set in
.I maskpri
are rejected.
The mask for an individual priority
.I pri
is calculated by the macro LOG_MASK(\fIpri\fP);
the mask for all priorities up to and including
.I toppri
is given by the macro LOG_UPTO(\fItoppri\fP).
The default allows all priorities to be logged.
.SH EXAMPLES
.nf
syslog(LOG_ALERT, "who: internal error 23");

openlog("ftpd", LOG_PID, LOG_DAEMON);
setlogmask(LOG_UPTO(LOG_ERR));
syslog(LOG_INFO, "Connection from host %d", CallingHost);

syslog(LOG_INFO|LOG_LOCAL2, "foobar error: %m");
.fi
.SH "SEE ALSO"
logger(1),
syslogd(8)


sieve/NEWS
$Id: NEWS,v 1.4 2000/05/17 15:47:06 ken3 Exp $

CMU Sieve 2.0
-------------

- Compliant with draft-showalter-sieve-11.txt and 
  draft-showalter-sieve-vacation-03.txt.

- Added support for the regex, imapflags, notify and subaddress extensions.
  See README for references.

- Verifies email addresses in redirect and vacation actions are syntactically
  correct (compliant with RFC822).

- Run-time error reporting.

- Changed callback interface to use callback contexts instead of individual
  parameters.  Also added an error string buffer for run-time error reporting.

- Vacation will not reply to any message containing an "auto-submitted"
  header containing anything other than "no".

CMU Sieve 1.4
-------------

Now included with imapd distribution (hell, why not?).

Error returning and recovering:
	added error recovering to the parser (but not much!)
	added error messages to the parser

Working on error returning and error recovering.
	run-time errors
	detect some errors in lexer?

Working on even better parsing:
	verify addresses could be addresses
	verify mailboxes could be mailboxes
	verify outgoing headers can be headers

CMU Sieve 1.3
-------------

Changed for integration with cyrus deliver.

CMU Sieve 1.2
-------------

Added additional callbacks (ok, so I want to make my integration with deliver
easier) and envelope and vacation support.

Made it compile without libcyrus.
It should compile without libcyrus, but then it does not implement the
"address" test.	 That's just too much work to do when I have a neato
library to do it for me.

Todo:
- regex matching

CMU Sieve 1.1
-------------

- Updated to draft-showalter-sieve-07bis.txt

- Simple API (see sieve_interface.h; currently mostly undocumented)

- Implements all of the optional features except "envelope"

- Maintains "if it parses, it probably runs" behavior. (Goal: minimize
  run-time errors.)

CMU Sieve 1.0
-------------

- prototype implementation


syslog/syslog.c
/*
 * Copyright (c) 1983, 1988, 1993
 *      The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *      This product includes software developed by the University of
 *      California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if defined(LIBC_SCCS) && !defined(lint)
static char sccsid[] = "@(#)syslog.c    8.8 (Berkeley) 8/30/95";
#endif /* LIBC_SCCS and not lint */

#include <sys/param.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/uio.h>
#include <syslog.h>
#include <netdb.h>

#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#ifdef BSD4_4
# include <paths.h>
#else
# ifndef _PATH_CONSOLE
#  define _PATH_CONSOLE "/dev/console"
# endif
# ifndef _PATH_LOG
#  define _PATH_LOG     "/dev/log"
# endif
#endif

#if __STDC__
# include <stdarg.h>
#else
# include <varargs.h>
# define const  /* */
#endif

#ifndef USESNPRINTF
# if defined(BSD4_4)
#  define USESNPRINTF   1       /* has snprintf(3), vsnprintf(3), etc. */
# else
#  define USESNPRINTF   0       /* cheesy old C library */
# endif
#endif

#ifndef LOG_PRI
# define LOG_PRI(p)     ((p) & LOG_PRIMASK)
#endif

#ifndef LOG_PERROR
# define LOG_PERROR     0
#endif

#define BUFSLOP         1024    /* overflow space */

static int      LogFile = -1;           /* fd for log */
static int      connected;              /* have done connect */
static int      LogStat = 0;            /* status bits, set by openlog() */
static const char *LogTag = NULL;       /* string to tag the entry with */
static int      LogFacility = LOG_USER; /* default facility code */
static int      LogMask = 0xff;         /* mask of priorities to be logged */
#if defined(BSD4_4)
extern char     *__progname;            /* Program name, from crt0. */
#else
char            *__progname = NULL;
#endif

/*
 * syslog, vsyslog --
 *      print message on log file; output is intended for syslogd(8).
 */
void
#if __STDC__
syslog(int pri, const char *fmt, ...)
#else
syslog(pri, fmt, va_alist)
        int pri;
        char *fmt;
        va_dcl
#endif
{
        va_list ap;
        extern void vsyslog();

#if __STDC__
        va_start(ap, fmt);
#else
        va_start(ap);
#endif
        vsyslog(pri, fmt, ap);
        va_end(ap);
}

void
vsyslog(pri, fmt, ap)
        int pri;
        register const char *fmt;
        va_list ap;
{
        register int cnt;
        register char ch, *p, *t;
        time_t now;
        int fd, saved_errno;
        int panic = 0;
        static int maxsend = BUFSIZ;
        char *stdp, fmt_cpy[1024], tbuf[BUFSIZ + BUFSLOP];
        extern void openlog();

#define INTERNALLOG     LOG_ERR|LOG_CONS|LOG_PERROR|LOG_PID
        /* Check for invalid bits. */
        if (pri & ~(LOG_PRIMASK|LOG_FACMASK)) {
                syslog(INTERNALLOG,
                    "syslog: unknown facility/priority: %x", pri);
                pri &= LOG_PRIMASK|LOG_FACMASK;
        }

        /* Check priority against setlogmask values. */
        if (!(LOG_MASK(LOG_PRI(pri)) & LogMask))
                return;

        saved_errno = errno;

        /* Set default facility if none specified. */
        if ((pri & LOG_FACMASK) == 0)
                pri |= LogFacility;

        /* Get connected. */
        if (!connected)
                openlog(LogTag, LogStat | LOG_NDELAY, 0);

        /* Build the message. */
        (void)time(&now);
        sprintf(tbuf, "<%d>", pri);
        p = tbuf + strlen(tbuf);
        strftime(p, sizeof (tbuf) - (p - tbuf), "%h %d %T ", localtime(&now));
        p += strlen(p);
        stdp = p;
        if (LogTag == NULL)
                LogTag = __progname;
        if (LogTag != NULL) {
                sprintf(p, "%s", LogTag);
                p += strlen(p);
        }
        if (LogStat & LOG_PID) {
                sprintf(p, "[%d]", getpid());
                p += strlen(p);
        }
        if (LogTag != NULL) {
                *p++ = ':';
                *p++ = ' ';
        }

        /* Substitute error message for %m. */
        for (t = fmt_cpy; ch = *fmt; ++fmt)
                if (ch == '%' && fmt[1] == 'm') {
                        ++fmt;
                        sprintf(t, "%s", strerror(saved_errno));
                        t += strlen(t);
                } else
                        *t++ = ch;
        *t = '\0';

#if USESNPRINTF
        cnt = maxsend - (p - tbuf) + 1;
        p += vsnprintf(p, cnt, fmt_cpy, ap);
        cnt = p - tbuf;
#else
        vsprintf(p, fmt_cpy, ap);
        p += strlen(p);
        cnt = p - tbuf;
        if (cnt > sizeof tbuf) {
                /* Panic condition. */
                panic = 1;
        }
        if (cnt > maxsend)
                cnt = maxsend;
#endif

        /* Output to stderr if requested. */
        if (LogStat & LOG_PERROR) {
                struct iovec iov[2];
                register struct iovec *v = iov;

                v->iov_base = stdp;
                v->iov_len = cnt - (stdp - tbuf);
                ++v;
                v->iov_base = "\n";
                v->iov_len = 1;
                (void)writev(STDERR_FILENO, iov, 2);
        }

        /* Output the message to the local logger. */
        for (;;) {
                if (send(LogFile, tbuf, cnt, 0) >= 0)
                        goto done;
                if (errno != EMSGSIZE || maxsend <= 256)
                        break;

                /* Message was too large -- back it off. */
                do {
                        maxsend -= 128;
                } while (cnt < maxsend && maxsend > 256);
                cnt = maxsend;
        }

        /*
         * Output the message to the console; don't worry about blocking,
         * if console blocks everything will.  Make sure the error reported
         * is the one from the syslogd failure.
         */
        if (LogStat & LOG_CONS &&
            (fd = open(_PATH_CONSOLE, O_WRONLY, 0)) >= 0) {
                (void)strcat(tbuf, "\r\n");
                cnt += 2;
                p = strchr(tbuf, '>') + 1;
                (void)write(fd, p, cnt - (p - tbuf));
                (void)close(fd);
        }

  done:
#if !USESNPRINTF
        /*
         * If we had a buffer overrun, log a panic and abort.
         * We can't return because our stack is probably toast.
         */
        if (panic) {
                syslog(LOG_EMERG, "SYSLOG BUFFER OVERRUN -- EXITING");
                abort();
        }
#endif
}

static struct sockaddr SyslogAddr;      /* AF_UNIX address of local logger */

void
openlog(ident, logstat, logfac)
        const char *ident;
        int logstat, logfac;
{
        if (ident != NULL)
                LogTag = ident;
        LogStat = logstat;
        if (logfac != 0 && (logfac &~ LOG_FACMASK) == 0)
                LogFacility = logfac;

        if (LogFile == -1) {
                SyslogAddr.sa_family = AF_UNIX;
                (void)strncpy(SyslogAddr.sa_data, _PATH_LOG,
                    sizeof(SyslogAddr.sa_data));
                if (LogStat & LOG_NDELAY) {
                        if ((LogFile = socket(AF_UNIX, SOCK_DGRAM, 0)) == -1)
                                return;
                        (void)fcntl(LogFile, F_SETFD, 1);
                }
        }
        if (LogFile != -1 && !connected)
                if (connect(LogFile, &SyslogAddr, sizeof(SyslogAddr)) == -1) {
                        (void)close(LogFile);
                        LogFile = -1;
                } else
                        connected = 1;
}

void
closelog()
{
        (void)close(LogFile);
        LogFile = -1;
        connected = 0;
}

/* setlogmask -- set the log mask level */
int
setlogmask(pmask)
        int pmask;
{
        int omask;

        omask = LogMask;
        if (pmask != 0)
                LogMask = pmask;
        return (omask);
}

